## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [131]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [132]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [133]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [134]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [135]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [136]:
# For the category with the most entries, which subcategory had the most entries?
df[df['category'] == 'consumables']['subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [137]:
# Which five clients had the most entries in the data?
df[['client_id']].value_counts(ascending=False).head(5)

client_id
33615        220
66037        211
46820        209
38378        207
24741        207
Name: count, dtype: int64

In [138]:
# Which five clients had the most entries in the data?
clientsList = df['client_id'].value_counts().head(5).index.to_list()
print(clientsList)

[33615, 66037, 46820, 38378, 24741]


In [139]:
# How many total units (the qty column) did the client with the most entries order order?
df[df['client_id'] == clientsList[0]]['qty'].sum()



64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [140]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
df[['unit_price', 'qty', 'line_subtotal']].head(2)


unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [141]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = df.apply(lambda x: x['unit_weight'] * x['qty'] * 7 if x['unit_weight'] * x['qty'] > 50 else x['unit_weight'] * x['qty'] * 10, axis=1)
df[['unit_weight', 'qty','shipping_price']].head(3)

unit_weight  qty  shipping_price
0         7.50  105         5512.50
1         1.49   21          312.90
2         1.68   39          458.64

In [142]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['line_price'] = df.apply(lambda x: round(((x['line_subtotal'] + x['shipping_price']) * 1.0925), 2), axis=1)
df[['line_price', 'shipping_price', 'line_subtotal']]

line_price  shipping_price  line_subtotal
0       131839.08         5512.50      115164.00
1          914.26          312.90         523.95
2         1077.12          458.64         527.28
3         1543.57          356.70        1056.18
4        11364.40         6500.20        3902.00
...           ...             ...            ...
54634     3564.87          519.75        2743.29
54635    14813.24         3849.30        9709.73
54636    49184.90        13832.00       31188.50
54637    15632.71        14143.36         165.76
54638    19711.43        14939.19        3103.31

[54639 rows x 3 columns]

In [143]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = df.apply(lambda x: round(((x['unit_cost'] * x['qty']) + x['shipping_price']), 2), axis=1)
df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ...         subcategory  \
0   8953482  2023-04-28          17        2023  ...            wall art   
1   8069089  2023-05-19          20        2023  ...                pens   
2   1902144  2023-01-29           4        2023  ...  project management   
3   9031802  2023-04-25          17        2023  ...                pens   
4   1322274  2023-05-28          21        2023  ...                misc   
5   1327730  2023-05-02          18        2023  ...              plants   
6   8080280  2023-04-19          16        2023  ...    kitchen supplies   
7   8946782  2023-02-14           7        2023  ...                pens   
8   9394118  2023-05-13          19        2023  ...      paper products   
9   9665852  2023-04-14          15        2023  ...             seating   

  unit_price unit_cost  unit_weight  qty  line_number  line_subtotal  \
0    1096.80    762.71         7.50  105            1      115164.00   
1      24.95     15.09         1.49   21            0         523.95   
2      13.52      7.86         1.68   39            6         527.28   
3      36.42     24.85         1.23   29            3        1056.18   
4     195.10    108.17        46.43   20            1        3902.00   
5     172.84    154.44         2.79   78            0       13481.52   
6     245.56    231.95         1.73   56            5       13751.36   
7      20.99     16.97         5.12   22            4         461.78   
8      83.86     59.00         0.27   49            3        4109.14   
9      66.88     54.15         2.47   78            1        5216.64   

   shipping_price  line_price  line_cost  
0         5512.50   131839.08   85597.05  
1          312.90      914.26     629.79  
2          458.64     1077.12     765.18  
3          356.70     1543.57    1077.35  
4         6500.20    11364.40    8663.60  
5         1523.34    16392.81   13569.66  
6          678.16    15764.25   13667.36  
7          788.48     1365.91    1161.82  
8          132.30     4633.77    3023.30  
9         1348.62     7172.55    5572.32  

[10 rows x 22 columns]

In [144]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df.apply(lambda x: (x['line_price'] - x['line_cost']), axis=1)
df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ... unit_price unit_cost  \
0   8953482  2023-04-28          17        2023  ...    1096.80    762.71   
1   8069089  2023-05-19          20        2023  ...      24.95     15.09   
2   1902144  2023-01-29           4        2023  ...      13.52      7.86   
3   9031802  2023-04-25          17        2023  ...      36.42     24.85   
4   1322274  2023-05-28          21        2023  ...     195.10    108.17   
5   1327730  2023-05-02          18        2023  ...     172.84    154.44   
6   8080280  2023-04-19          16        2023  ...     245.56    231.95   
7   8946782  2023-02-14           7        2023  ...      20.99     16.97   
8   9394118  2023-05-13          19        2023  ...      83.86     59.00   
9   9665852  2023-04-14          15        2023  ...      66.88     54.15   

  unit_weight  qty  line_number  line_subtotal  shipping_price  line_price  \
0        7.50  105            1      115164.00         5512.50   131839.08   
1        1.49   21            0         523.95          312.90      914.26   
2        1.68   39            6         527.28          458.64     1077.12   
3        1.23   29            3        1056.18          356.70     1543.57   
4       46.43   20            1        3902.00         6500.20    11364.40   
5        2.79   78            0       13481.52         1523.34    16392.81   
6        1.73   56            5       13751.36          678.16    15764.25   
7        5.12   22            4         461.78          788.48     1365.91   
8        0.27   49            3        4109.14          132.30     4633.77   
9        2.47   78            1        5216.64         1348.62     7172.55   

   line_cost  line_profit  
0   85597.05     46242.03  
1     629.79       284.47  
2     765.18       311.94  
3    1077.35       466.22  
4    8663.60      2700.80  
5   13569.66      2823.15  
6   13667.36      2096.89  
7    1161.82       204.09  
8    3023.30      1610.47  
9    5572.32      1600.23  

[10 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [145]:
# Check your work using the totals above
order2742071 = round(df[df['order_id'] == 2742071][['order_id', 'line_price']].sum()['line_price'], 2)
order2173913 = round(df[df['order_id'] == 2173913][['order_id', 'line_price']].sum()['line_price'], 2)
order6128929 = round(df[df['order_id'] == 6128929][['order_id', 'line_price']].sum()['line_price'], 2)

print(f'Order 2742071 total price: ${order2742071:,.2f}')
print(f'Order 2173913 total price: ${order2173913:,.2f}')
print(f'Order 6128929 total price: ${order6128929:,.2f}')

Order 2742071 total price: $152,811.89
Order 2173913 total price: $162,388.71
Order 6128929 total price: $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [146]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# Get total spending (line_price) for each of the top 5 clients by quantity
top_5_spending = df[df['client_id'].isin(clientsList)][['client_id', 'line_price']]
top_5_spending = top_5_spending.groupby('client_id').sum()['line_price']
top_5_spending



client_id
24741    82268892.04
33615     8377308.53
38378    12906550.87
46820     9743794.33
66037    10259514.80
Name: line_price, dtype: float64

In [147]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
# Filter for top 5 clients and select relevant columns
top_5_summary = df[df['client_id'].isin(clientsList)].groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum', 
    'line_price': 'sum',
    'line_cost': 'sum',
    'line_profit': 'sum'
}).reset_index()
top_5_summary

client_id     qty  shipping_price   line_price    line_cost  line_profit
0      24741  239862      5126448.37  82268892.04  45688899.71  36579992.33
1      33615   64313      1828984.89   8377308.53   6175313.91   2201994.62
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
3      46820   75768      1601448.84   9743794.33   7007191.64   2736602.69
4      66037   43018      1395151.85  10259514.80   7004482.98   3255031.82

In [148]:
# Format the data and rename the columns to names suitable for presentation.
# Rename columns to presentation format
top_5_summary = top_5_summary.rename(columns={
    'client_id': 'Client ID',
    'qty': 'Units',
    'shipping_price': 'Shipping',
    'line_price': 'Total Revenue', 
    'line_cost': 'Total Cost',
    'line_profit': 'Total Profit'
})

# Define the money columns. 
money_columns = ['Shipping', 'Total Revenue', 'Total Cost', 'Total Profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(amount):
    """Convert a dollar amount to millions."""
    return amount / 1_000_000


# Apply the currency_format_millions function to only the money columns. 
# Apply currency formatting to money columns
for col in money_columns:
    top_5_summary[col] = top_5_summary[col].apply(currency_format_millions)


# Rename the columns to reflect the change in the money format. 
top_5_summary = top_5_summary.rename(columns={
    'Shipping': 'Shipping (Millions)',
    'Total Revenue': 'Total Revenue (Millions)',
    'Total Cost': 'Total Cost (Millions)',
    'Total Profit': 'Total Profit (Millions)'
    })

top_5_summary

Client ID   Units  Shipping (Millions)  Total Revenue (Millions)  \
0      24741  239862             5.126448                 82.268892   
1      33615   64313             1.828985                  8.377309   
2      38378   73667             3.429455                 12.906551   
3      46820   75768             1.601449                  9.743794   
4      66037   43018             1.395152                 10.259515   

   Total Cost (Millions)  Total Profit (Millions)  
0              45.688900                36.579992  
1               6.175314                 2.201995  
2               9.634721                 3.271830  
3               7.007192                 2.736603  
4               7.004483                 3.255032

In [149]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
top_5_summary_sorted = top_5_summary.sort_values('Total Profit (Millions)', ascending=False)
top_5_summary_sorted


Client ID   Units  Shipping (Millions)  Total Revenue (Millions)  \
0      24741  239862             5.126448                 82.268892   
2      38378   73667             3.429455                 12.906551   
4      66037   43018             1.395152                 10.259515   
3      46820   75768             1.601449                  9.743794   
1      33615   64313             1.828985                  8.377309   

   Total Cost (Millions)  Total Profit (Millions)  
0              45.688900                36.579992  
2               9.634721                 3.271830  
4               7.004483                 3.255032  
3               7.007192                 2.736603  
1               6.175314                 2.201995